src/Billboard/Billboard.sol

```diff
@@ -0,0 +1,165 @@
+//SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+import "./BillboardAuction.sol";
+import "./BillboardRegistry.sol";
+import "./IBillboard.sol";
+import "./IBillboardAuction.sol";
+import "./IBillboardRegistry.sol";
+
+contract Billboard is IBillboard {
+    address public admin;
+
+    BillboardAuction public auction;
+
+    BillboardRegistry public registry;
+
+    constructor() {
+        admin = msg.sender;
+    }
+
+    //////////////////////////////
+    /// Modifiers
+    //////////////////////////////
+
+    modifier isValidAddress(address value_) {
+        if (value_ == address(0)) {
+            revert InvalidAddress();
+        }
+        _;
+    }
+
+    modifier isAdmin(address value_) {
+        if (admin == address(0)) {
+            revert AdminNotFound();
+        }
+        if (value_ == address(0)) {
+            revert InvalidAddress();
+        }
+        if (value_ != admin) {
+            revert Unauthorized("admin");
+        }
+        _;
+    }
+
+    //////////////////////////////
+    /// Upgradability
+    //////////////////////////////
+
+    /// @inheritdoc IBillboard
+    function upgradeAuction(address contract_) external isValidAddress(contract_) isAdmin(msg.sender) {
+        auction = BillboardAuction(contract_);
+    }
+
+    /// @inheritdoc IBillboard
+    function upgradeRegistry(address contract_) external isValidAddress(contract_) isAdmin(msg.sender) {
+        registry = BillboardRegistry(contract_);
+    }
+
+    /// @inheritdoc IBillboard
+    function setIsOpened(bool value_) external isAdmin(msg.sender) {
+        if (address(registry) == address(0)) {
```

Comment:
L61-L66 can use `isValidAddress` instead?

Reply:
or we don't need L61-L66 actually since the transaction will be failed if registry or auction are zero addresses.

Reply:
You're right tx will be failed.